// @prisma-cms/e-commerce/src/server/modules/product/index.js

import chalk from "chalk";

import { fileLoader, mergeTypes } from 'merge-graphql-schemas';

import PrismaModule from "@prisma-cms/prisma-module";


class ECommerceModule extends PrismaModule {


  // constructor(options = {}) {

  //   let {
  //     modules = [],
  //   } = options;

  //   modules = modules.concat([
  //   ]);

  //   Object.assign(options, {
  //     modules,
  //   });

  //   super(options);

  // }


  getSchema(types = []) {

    let schema = fileLoader(__dirname + '/schema/database/', {
      recursive: true,
    });

    if (schema) {
      types = types.concat(schema);
    }

    let typesArray = super.getSchema(types);

    return typesArray;

  }


  getApiSchema(types = []) {


    let apiSchema = super.getApiSchema(types, []);


    let schema = fileLoader(__dirname + '/schema/api/', {
      recursive: true,
    });

    apiSchema = mergeTypes([apiSchema.concat(schema)], { all: true });

    return apiSchema;

  }


  // getExcludableApiTypes(){

  //   return super.getExcludableApiTypes([
  //   ]);

  // }


  getResolvers() {

    console.log("getResolvers");

    const resolvers = super.getResolvers();

    Object.assign(resolvers.Query, {
      products: this.products,
    });


    Object.assign(resolvers.Mutation, {
    });


    Object.assign(resolvers, {
    });


    return resolvers;
  }


  products(source, args, ctx, info) {

    console.log("products args", args);

    return [
      {
        __typename: "ProductType1",
        id: 234,
        name: "DSfsdf",
      },
      {
        __typename: "ProductType2",
        id: 234,
        name: "FDGhfgdhd",
      },
      {
        __typename: "ProductType3",
        id: 2434,
        name: "rgSDasd",
      },
    ];
  }


}


export default ECommerceModule;
